### ISSUE TITLE
Scrolling issues

### ISSUE THREAD

**Issue by Eifelschaf** [TonicArtos/SuperSLiM#55]
Hi,

I think I found another issue:

I have a grid layout with variable size cards. If the first card in the last line is bigger than the rest it is cut off. Also I can't scroll back to the top in this case.

The layout item is completely displayed if I first scroll down and then turn the display (probably repainting the view)

Edit: I am using the width-based version of the grid layout.
Edit 2: If I scroll fast enough I will get the complete item visible

**Comment by PatrickDattilio** [comment]
This sounds like a StaggeredGridView, which is not yet supported.

**Comment by Eifelschaf** [comment]
It sounds like one, but it is not ;) Due to the fact that staggered layouts are not yet supported I put the views in a grid. It works perfectly fine if the biggest item is not the first one in a grid row, so there must be some small measuring issue. I am trying to find it right now, but it's quite hard if you don't know where to look.

**Comment by TonicArtos** [comment]
Grid layout does handle different height items but makes the row fit the max row item height. I am not seeing this error in v0.4.4.

**Comment by Eifelschaf** [comment]
I am, though (also using 0.4.4). How can I help you to reproduce this? Unfortunately I can not provide the project itself.

**Comment by Eifelschaf** [comment]
Where's the max row height calculated? Perhaps I can fix it myself.

**Comment by TonicArtos** [comment]
GridSLM#fillRow(…)

**Comment by TonicArtos** [comment]
However, there seems to be a problem passing through margins in the layout params.

**Comment by Eifelschaf** [comment]
Even if I remove the margins I can not scroll back to the top any more. I will have a look at this tomorrow if you can't reproduce it.

**Comment by TonicArtos** [comment]
Okay the margin issue is just a red herring, the margins were being stripped as they were passed through the grid layout layout param constructor. 

Anyway, there is actually something strange going on with CardView. Sometimes cards are not measuring as the same size.

**Comment by TonicArtos** [comment]
Is the card view completely cut off? Or is text content cut off? If it is the latter, then I think it is basically the same issue as #56. I suggest making sure you are using different view type ids for views going to different layouts, and for headers too.

**Comment by Eifelschaf** [comment]
The card view is completely cut off. I am also using different view types, but the views themselves are of a dynamic height. I will display some dummy values and post a few screen shots later.

**Comment by Eifelschaf** [comment]
I just hit the autocompletion a few times to fill some text views. Here are my results:

![scroll_works](https://cloud.githubusercontent.com/assets/1673940/6521338/e5cb5802-c3d6-11e4-84c8-c4ed9096bbaa.png)
If the second card is bigger than the first one everythingn is ok.

![scroll_broken](https://cloud.githubusercontent.com/assets/1673940/6521340/ea6336f0-c3d6-11e4-83a1-a1d902396d13.png)
However if the first one is bigger the recycler view doe snot scroll any more and cards may be cut off if I scroll there slowly. If I scroll to the position with a fast movement I see the complete card (as in the image).

**Comment by TonicArtos** [comment]
Are those TextEdit views inside the cards? Maybe that is related.

**Comment by Eifelschaf** [comment]
Yes, but the issue is not related. They were just the only ones I could fill easily. 
I set the values by typing into a dialog and then call settext on the EditText later.

The issue also happens with LinearLayouts that I fill programmatically (that was the original reason). It really seems to depend on the size of the views.

**Comment by Eifelschaf** [comment]
I also get weird values from getDecoratedBottom(). They seem to be changing based on the scroll speed.

**Comment by Eifelschaf** [comment]
I think I found the issue that blocks the scrolling, but I don't think that this was the original problem...

Look at this commit in my fork, please: https://github.com/Eifelschaf/SuperSLiM/commit/90f6c8b4c0e9f203554a7874fe3c607f55c54325

Does this make sense? This change allows me to scroll up at least (still with the cut card)

**Comment by TonicArtos** [comment]
If that early return is the issue then it means the section bottom edge is not being calculated properly.

**Comment by TonicArtos** [comment]
Hang on, is this only happening for the very last item in the entire (super) layout, not just a grid layout?

**Comment by TonicArtos** [comment]
Aha, there we go.

**Comment by Eifelschaf** [comment]
Yes, I think I nailed it down. The getAnchorAtEnd() return the last view, not the biggest in the last row :)

**Comment by TonicArtos** [comment]
The reason for the cut off is that the grid is not computing the bottom edge correctly, so the scroll is not advancing to the real end of the layout.

**Comment by TonicArtos** [comment]
Yeah, getAnchorAtEnd should be passing through to the SLM.

**Comment by Eifelschaf** [comment]
Do you think it's a big deal to get this fixed?

**Comment by TonicArtos** [comment]
Actually, it should be using getAnchorAtEnd, it should be asking the SLM directly for the bottom edge.

**Comment by TonicArtos** [comment]
It should be a minor fix. Give me a couple of minutes and I'll push a snapshot for you to test.

**Comment by TonicArtos** [comment]
Yeah, it was trivial to fix. I am just testing it now. There are some other small fixes that have been pulled together with this and I want to check it is all working together.

**Comment by Eifelschaf** [comment]
Sure, take your time.

**Comment by TonicArtos** [comment]
Okay, I didn't come across any problems. Check the snapshot, if it is fine I'll push up the new version.

```groovy
repositories {
    maven { url "https://oss.sonatype.org/content/repositories/snapshots/" }
}
```

```groovy
dependencies {
    compile 'com.tonicartos:superslim:0.4.5-SNAPSHOT'
}﻿
```

**Comment by TonicArtos** [comment]
sorry

**Comment by TonicArtos** [comment]
hangon

**Comment by TonicArtos** [comment]
Okay, I actually merged the fix in to the snapshot this time. You can test it now :)

**Comment by Eifelschaf** [comment]
You are awesome! It works perfectly fine now.

**Comment by Eifelschaf** [comment]
stop.. I just encounteredit again. I have to re-check this...

**Comment by Eifelschaf** [comment]
OK, the scrolling is fixed, but the card is still cut if I scroll slowly. Are you sure you forward the measurement calls everywhere?

**Comment by TonicArtos** [comment]
Yeah, I am seeing the bug in master after merging it, but not on the release branch. There is something I have to check.

**Comment by Eifelschaf** [comment]
As I said: take your time :)

**Comment by TonicArtos** [comment]
Oh, hey, can you make sure to refresh the snapshot? Because there were two copies released close to each other, the unfixed one might be cached on your machine. Just to be sure.

**Comment by Eifelschaf** [comment]
I only got it after the second release, but I will check again.

2015-03-06 11:32 GMT+01:00 Tonic Artos <notifications@github.com>:

**Comment by TonicArtos** [comment]
I am definitely not seeing the bug anymore. I fixed the code divergence on master too.

**Comment by Eifelschaf** [comment]
I just did a full rebuild with 0.4.5-snapshot and it's still there.. did you try scrolling slowly?

**Comment by TonicArtos** [comment]
Yeah, slow scrolling is fine here. I've tested with a number of views including EditText. I just noticed couple of bugs with EditText, but they are quite different (performance and keyboard interaction).

**Comment by Eifelschaf** [comment]
I will quickly try a rebuild with the current source code. It's the early_release_4 branch, right?

**Comment by TonicArtos** [comment]
Yes

**Comment by Eifelschaf** [comment]
Nope, still there, even with the latest source. I will have a look at it now and let you know if I find something more specific.

**Comment by Eifelschaf** [comment]
As I thought: the scrollVerticallyBy method still uses the wrong view for measurement

**Comment by Eifelschaf** [comment]
ahhh shit.. This may have been my fault...

**Comment by TonicArtos** [comment]
Okay. Here is what should be happening. On scroll a leading edge is calculated. For a scroll towards the end of the content that edge is the height of the view + the scroll displacement. 

To fill the leading are in the scroll, the gridSlm should be measuring and laying out content row by row, up to that edge. The line (markerLine) up to which the grid has filled should be the bottom of the last grid row laid out.

The fix to this problem was an adjustment to use the SLM to calculate the end of the content (grid slm knows about rows, layout manager does not), rather than assuming the anchor position is at the bottom edge.

**Comment by Eifelschaf** [comment]
OK. It's fixed. It was just some very weird behavior:

I removed all sectionmanager markups from my non-header views as they were ignored for layout anyways. Turns out that they are taken into account to get the slm for measurement now.

**Comment by TonicArtos** [comment]
and in v0.5.0 they don't matter again

**Comment by TonicArtos** [comment]
Actually, they were not quite always ignored. For the scroll indicator they are required in v0.4.x.

**Comment by Eifelschaf** [comment]
great! Thanks! I will go with the current solution for now (multiple markup) and wait for 0.5 to change them.

**Comment by TonicArtos** [comment]
Yeah, sorry. It is a pain, I know. :(

**Comment by Eifelschaf** [comment]
It's ok as long as it works :) Thanks for helping me.

**Comment by TonicArtos** [comment]
But now I have two more problems to fix for EditText.

**Comment by TonicArtos** [comment]
:) You're welcome, reporting bugs helps me too.

**Comment by Eifelschaf** [comment]
If i can help you with the edit text just let me know.

**Comment by TonicArtos** [comment]
Thanks for offering. I am going to tackle it after #53. I will have to make a number of changes to the layout code so I might be able to fix it with that anyway.

**Comment by TonicArtos** [comment]
Alright, v0.4.5 should be available now.
